File: lim/그리디/상태 반전이 가능한 문제 - 상하좌우 반전시키기.py
import sys


def sol(n, grid):
    drs = [1, -1, 0, 0, 0]
    dcs = [0, 0, 1, -1, 0]

    def in_range(r, c):
        return r >= 0 and c >= 0 and r < n and c < n

    def flip(r, c):
        for dr, dc in zip(drs, dcs):
            nr, nc = r + dr, c + dc
            if in_range(nr, nc):
                grid[nr][nc] = (grid[nr][nc] + 1) % 2

    count = 0
    for i in range(1, n):
        for j in range(n):
            if grid[i-1][j] == 0:
                flip(i, j)
                count += 1

    for i in range(n):
        if grid[-1][i] == 0:
            print(-1)
            return
    print(count)


n = int(sys.stdin.readline())
grid = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
sol(n, grid)

File: lim/그리디/상태_반전이_가능한_문제_상하좌우_반전시키기.java
package lim.그리디;

import java.util.*;
import java.io.*;

public class 상태_반전이_가능한_문제_상하좌우_반전시키기 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n;
    static int[] drs = { 1, 0, -1, 0 }, dcs = { 0, 1, 0, -1 };
    static int[][] grid;

    static boolean inRange(int r, int c) {
        return r >= 0 && c >= 0 && r < grid.length && c < grid[0].length;
    }

    static void flip(int r, int c) {
        grid[r][c] ^= 1;
        for (int i = 0; i < drs.length; i++) {
            int nr = r + drs[i];
            int nc = c + dcs[i];
            if (!inRange(nr, nc)) {
                continue;
            }
            grid[nr][nc] ^= 1;
        }
    }

    static void sol() throws IOException {
        int flipCount = 0;
        for (int i = 1; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i - 1][j] == 0) {
                    flip(i, j);
                    flipCount++;
                }
            }
        }
        for (int i = 0; i < grid[0].length; i++) {
            if (grid[grid.length - 1][i] == 0) {
                bw.write("-1");
                return;
            }
        }
        bw.write(flipCount + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        grid = new int[n][n];
        for (int i = 0; i < grid.length; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < grid[0].length; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}